### ISSUE TITLE
Make hashtable literals use type `System.Management.Automation.OrderedHashtable`

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#19073]
### Summary of the new feature / enhancement

Since 7.3,  a new, `[hashtable]`-derived [`System.Management.Automation.OrderedHashtable`](https://learn.microsoft.com/en-US/dotnet/api/System.Management.Automation.OrderedHashtable) type is now available, which provides a backward-compatible implementation that preserves _entry-definition order_.

As of this writing (as far as I'm aware of), this type is only used with `ConvertFrom-Json -AsHashtable`.
It would equally be useful in `ConvertFrom-StringData` and `Import-PowerShellDataFile` (and potentially other `[hashtable]`-outputting cmdlets) - see #19070.

Given that the backward-compatibility impact is so low - as was deemed to be the case in the context of `ConvertFrom-Json -AsHashtable`, see https://github.com/PowerShell/PowerShell/issues/18524#issuecomment-1314400113 - also making hashtable _literals_ use `System.Management.Automation.OrderedHashtable` seems like a beneficial change, given that:

* maintaining entry-definition order is _always_ beneficial, at the very least to the _human_ observer
* the performance impact of maintaining this order is negligible.

```powershell
# WISHFUL THINKING

PS> @{ foo = 1; bar = 2; zulu = 3; midi = 4 }

Name                           Value  # ENTRY-DEFINITION ORDER PRESERVED
----                           -----
foo                            1
bar                            2
zulu                           3
midi                           4


PS> @{ foo = 1; bar = 2; zulu = 3; midi = 4 }.GetType().FullName
System.Management.Automation.OrderedHashtable
```

That is, hashtable literals will then be _implicitly_ ordered, without needing an `[ordered]` pseudo cast.

However, unlike with `[ordered]` the resulting hashtable will _not_ support _direct_ positional indexing, though it will be available via the `.Values` collection; that is, the following will be equivalent:

```powershell
PS> ([ordered] @{ foo = 1; bar = 2; zulu = 3; midi = 4 })[2]
3

# WISHFUL THINKING
PS> @{ foo = 1; bar = 2; zulu = 3; midi = 4 }.Values[2]
3
```

**Comment by Jaykul** [comment]
You would break the world. 
Every script that calls an API (like, for instance, a PowerShell cmdlet) that takes a hashtable as a parameter, would just crash instead.

Then, to try and fix it, you'd put in implicit casting from IDictionary to hashtables, and you'd break the world even more...

**Comment by Jaykul** [comment]
This would have all sorts of second order effects you can't predict, like changing the order of splatting parameter resolution, and inflicting potentially huge memory load on scripts that were using simple hashtables with "lots" of items in them.

In my opinion, not worth the breakage, since it doesn't really solve any _problems_.

**Comment by mklement0** [comment]
> like changing the order of splatting parameter resolution

With a `[hashtable]`, there's currently no way to control the order of parameter binding via splatting - you won't know the order of the entries ahead of time.
Are you saying that the order _matters_? Clearly, it currently doesn't. Why should it suddenly matter if you get _predictable_ ordering?

> inflicting potentially huge memory load on scripts that were using simple hashtables with "lots" of items in them

I have no data to back that up, but that doesn't strike me as a real-world concern.
Clearly, it was deemed not to be a concern for  `ConvertFrom-Json -AsHashtable` - despite the open-ended nature of the JSON input with respect to its size.

> since it doesn't really solve any problems.

It solves the same problem as it does in the context of `ConvertFrom-Json -AsHashtable`, only _fundamentally_. 
It is always somewhat disconcerting _to the human observer_ to see the entries jumbled the way they currently are, even if the lack of predictable ordering doesn't matter _technically_.
Fixing this problem once and for all seems well worth it to me.

**Comment by doctordns** [comment]
I am not convinced there is a real problem here. Non-Ordered hash tables are a little bit faster, and that performance is by default.  Having used PowerShell for a while, my use of [ordered] is pretty rare - and only when I need to ensure the order (e.g. when creating a custom object from a hash table). 

I feel the current default, which probably accounts for the VAST majority of the use case, is a good one. Were we designing PowerShell from scratch (or were this 2002), I might agree - but this boat has sailed.

**Comment by mklement0** [comment]
@doctordns, there was never a _problem_ here - only a _proposal_ to make hashtable literals _user-friendly_, by making them work as they arguably always should have.

My _hope_ (now somewhat dashed - see below) - now we have a `[hashtable]`-_derived_ type that preserves input order -was that we can use it _by default_, with _little_ risk of breaking existing code (a risk that was already deemed acceptable in the context of `ConvertFrom-Json -AsHashtable`) and with little performance impact, while reaping all the benefits.

Unfortunately, benchmarks that I didn't run until now tell me that while _accessing_ ordered hashtables is virtually as fast as unordered ones, _building_ them is substantially slower (plus increased memory usage, as @Jaykul  had pointed out, though I'd consider that alone less of a problem).

This is unlikely to matter for _literals_, but if you start with `@{}` and _programmatically_ build a hashtable with many entries, the impact is very noticeable:

Here is a benchmark of the 15-run average for creating and accessing 1000 entries (`$ht` = `[hashtable]`; `$ohtSma` = `[System.Management.Automation.OrderedHashtable]`; `$oht` = `[ordered]`); note the `Factor` column:

```none
VERBOSE: Creating...

Factor Secs (15-run avg.) Command                                      TimeSpan
------ ------------------ -------                                      --------
1.00   0.000              foreach($k in $Keys) { $ht[$k] = 'foo' }     00:00:00.0004864
6.49   0.003              foreach($k in $Keys) { $ohtSma[$k] = 'foo' } 00:00:00.0031577
7.38   0.004              foreach($k in $Keys) { $oht[$k] = 'foo' }    00:00:00.0035899

VERBOSE: Accessing...

Factor Secs (15-run avg.) Command                              TimeSpan
------ ------------------ -------                              --------
1.00   0.000              foreach($k in $Keys) { $ohtSma[$k] } 00:00:00.0004639
1.05   0.000              foreach($k in $Keys) { $oht[$k] }    00:00:00.0004861
1.05   0.000              foreach($k in $Keys) { $ht[$k] }     00:00:00.0004886
```

Therefore, I now see that the proposed change is problematic, and I'm closing it.

---

<details>
<summary>Benchmark source code (uses <a href="https://gist.github.com/mklement0/9e1f13978620b09ab2d15da5535d1b27"><code>Time-Command</code></a>:</summary>

```powershell
$keys = 1..1000 | % { -join ('a'..'z' | Get-Random -Count 5) }
$ht = @{}
$oht = [ordered] @{}
$ohtSma = [System.Management.Automation.OrderedHashtable]::new()

Write-Verbose -Verbose 'Creating...'
Time-Command -Count 15 @( 
  { foreach($k in $Keys) { $ht[$k] = 'foo' } }
  { foreach($k in $Keys) { $oht[$k] = 'foo' } }
  { foreach($k in $Keys) { $ohtSma[$k] = 'foo' } }
) | Out-Host

Write-Verbose -Verbose 'Accessing...'
Time-Command -Count 15 @( 
  { foreach($k in $Keys) { $ht[$k] } }
  { foreach($k in $Keys) { $oht[$k] } }
  { foreach($k in $Keys) { $ohtSma[$k] } }
) | Out-Host
```

</details>

**Comment by iRon7** [comment]
> This would have all sorts of second order effects you can't predict, like changing the order of splatting parameter resolution, and inflicting potentially huge memory load on scripts that were using simple hashtables with "lots" of items in them.

The side affects from these change are clearly unnoticed.
Knowing that somewhere along the line they have changed somewhere between  version `7.2.10` and `7.3.6`:

```PowerShell
PowerShell 7.2.10
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.

   A new PowerShell stable release is available: v7.3.9
   Upgrade now, or check out the release page at:
     https://aka.ms/PowerShell-Release?tag=v7.3.9

PS C:\Users\boder> @{ NonNodeData = 1; AllNodes = 2}

Name                           Value
----                           -----
AllNodes                       2
NonNodeData                    1
```

```PowerShell
PowerShell 7.3.9
PS C:\Users\boder> @{ NonNodeData = 1; AllNodes = 2}

Name                           Value
----                           -----
NonNodeData                    1
AllNodes                       2
```

See also: #13566

Meaning, only the performance argument still stands to not change this behavior.
Besides, a new `OrderedHashtable` might not break anything by itself but breaks (as far as I can determine) things where it replaces a formal hashtable, see: #19928
